### ISSUE TITLE
Motivations

### ISSUE THREAD

**Issue by Turbo87** [maggiepint/proposal-temporal#19]
As moment.js and js-joda demonstrate this proposal could be implemented as a library too. I think we should make the actual motivations a little more clear why this should be built into the browser/platform, and what the advantages are compared to being implemented as a third-party lib.

- `Date` is built-in too, but the general consensus seems to be that it's not a great design and has a lot of flaws. These flaws can not be fixed without breaking API unfortunately so to fix it, the whole class would need to be deprecated and replaced with something else. (Alternative is to keep the broken API forever and only rely on third-party libs for sane date/time code)

- The timezone database is large, and shipping it to the browser is costly. It would be nice if the browser/platform provided at least the timezone database natively so that libraries like moment.js or js-joda could take advantage of it. (Alternative is shipping it with the application code or figuring out what timezone info is needed and loading it async from the server, which requires a lot more complexity though)

The first point would probably favor the current proposal, while the second point would be resolved if only some low-level primitives would be offered by the platform. I personally haven't yet decided what road I'd choose, but it is clear that any solution would be better than the current state. :)

**Comment by rxaviers** [comment]
>  I think we should make the actual motivations a little more clear why this should be built into the browser/platform, and what the advantages are compared to being implemented as a third-party lib.

👍

**Comment by maggiepint** [comment]
This is a broad question, but the short version is that people expect working datetime code in any higher-level general purpose programming language, and JavaScript is certainly that.

I wrote some pretty extensive blog posts about the WHY of travelling this road here: https://maggiepint.com/2017/04/09/fixing-javascript-date-getting-started/

As far as Moment or js-Joda not just carrying the torch on this - well, I maintain Moment. We're struggling. We get a lot of flack for the size of the library - even without the locales or the timezone data included. If we're still considered big with those things removed, than all I can say is that the datetime problem domain is too big for a library - all of our code paths are used. People point to DateFNS and say that it's better because it's modular, but they actually are bigger than moment when built in entirety (at lest when last I checked), and they have fewer features. For instance, they don't have a date parser (10% of Moment'size). Since Date parsing in JS is totally broken, this is a serious concern.

So in short the motivations are:
1. The date object in JavaScript is broken so many ways it can't be fixed
2. People expect date handling code in any higher level language
3. The problem domain is extremely common and the code to support it exceeds the size of what people are willing to have in most libraries.

Does that cover this?

**Comment by ndaidong** [comment]
Why we need a complex as such API just for calculating date time. JavaScript native API should provide only the basic things, never goes deeply into detail. And I'm happy with the current Date object.

**Comment by snboisen** [comment]
@maggiepint I fully and enthusiastically agree with your stated motivations. Could they be put into the proposal, so that people can be reminded why this is a stellar proposal? 😃

**Comment by rxaviers** [comment]
> The date object in JavaScript is broken so many ways it can't be fixed

In your blog post, you mention **two** reasons: mutability & parser. You mention that anything else would be fixable.

In the motivation section, there's no mention about parser and no explanation about what's the efforts to handle immutability by a library in userland… I encourage you to enrich your motivation section by including the challenges this proposal is solving. See below for details.

> The problem domain is extremely common and the code to support it exceeds the size of what people are willing to have in most libraries.

Can you elaborate a bit more on this point please? The size issue is obvious to me in two places: **time zone** support & **non-gregorian** calendars support, but this proposal (as it reads now) doesn't fix non-gregorian calendars. Is there anything else but to avoid time zone data? Some numbers would be helpful.

Note that anything related to internationalization is handled (or should be) by Ecma-402, so I'm ignoring localization concerns in here.

## Overall

Is this proposal about *convenient features* or *must have features*? The time zone support is a clear must have one, but overall it's unclear to me the weight and importance of each of the several micro problems being solved here… It would be great if you could list them all in your motivation section and explain what's the effort to solve each in userland libraries.

For example, handling time zone support in userland requires a significant amount of data and therefore the benefit of having such support in native JavaScript is obvious on that regard. Handling non-gregorian calendars requires tons of crazy algebras and therefore the benefit of having such support in native JavaScript would be similarly obvious if solved by this proposal.

Immutability… Technically, a thin wrapper on top of existing Date is sufficient to provide such feature [1]. Am I missing something? If so, it would be great to explain that in the motivation section.

CivilDate, CivilTime, CivilDateTime… Technically, a thin wrapper on top of existing Date is sufficient to provide such feature [2]. Am I missing something? If so, it would be great to explain that in the motivation section.

Addition, subtraction, and other computation… Similarly, it's good to explain the challenges solved here. Convenience only or something else?

------

1:  A library can provide immutability by proxying the setter methods and cloning the passed date before actually setting it and then return that.

2: A library can use the native Date UTC methods under the hoods and provide time without any time zone reference.

**Comment by andreineculau** [comment]
FWIW I was also going to open an issue, if it wasn't for this one - for the simple fact that motivations contrived to "Date has been a long time pain point in ECMAScript." cannot possibly end with a positive outcome. For every single thing that exists in this world there's always a bunch that experience a pain point. Pain is subjective, not objective.

# Regarding the size argument.

For those that are old enough, it may not come as a surprise that in the days right after the AJAX hip times, JS devs started to worry about the sizes of the code they push, and more so about the fact that most of that code is not theirs, but supporting code - libraries (prototype), frameworks (ExtJS), etc.

An idea popped out on different channels, but died in ignorance, that browsers could act as a local CDN for the most popular libraries and frameworks. I even have a childish (since then, I graduated kindergarten...) post about it - http://blog.andreineculau.com/2006/05/19/future-of-web-browsers#content

Long story short, a proposal to implement an API for a local CDN would solve this, and so much more. But the argument of size alone stands true for a gazillion other libraries that different groups of devs consider their bread and butter and they too would love not to have to bundle them and have them available in all JS environments.

# Regarding design simplicity and the logic of the motivation.

Design, and even more so language design, is about getting it as simple as possible, but not simpler.

Did JS really need to add String.prototype.padStart() or String.prototype.padEnd() ? The [rationale](https://github.com/tc39/proposal-string-pad-start-end#rationale) there was also relative pain "more painful than it should be" , feelings in general "the language feels incomplete", probabilities "highly probable ... inefficient" and inaccuracies such as devs "no longer have to implement these common functions" --- as the infamous left-pad event has shown, devs didn't need to do that in the first place.

With that in mind, if the consensus is that the current Date is broken for x, y, z reasons, then the top action point would sound like "Deprecate Date. Using Date from now on will print a console warning ..." (just like we had "Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check http://xhr.spec.whatwg.org/.").

The second action point I was interested in was - what was lacking in the current Date. @mj1856 highlights such a detail that my company and I would very much applaud. Timezone support, better parsing, etc can all be done in a library. Increase timestamp precision, not possible.

Third action point would highlight how the new design actively skips all the broken things in the current design. For instance, https://youtu.be/aVuor-VAWTI?t=623 . If all that Date was doing in the beginning was give you back a unix timestamp, then nothing would have been broken. We would then discuss how we need timezone support. And then how we need locales. And each and every one of those points could be handled separately, in the same or different ways (i.e. localization is its own domain, where time is just added context; timezones are unrelated to localization).

Reading how momentjs gets "flack for the size" - myself a momentjs user aware of the situation - is unfortunately **not** what an language-design action point should tackle imo.

# Posterity.

Fast forward in time, whenever this proposal will not be a proposal anymore, people will rant once again how temporal is broken, but a fraction will actually have it in them to say "maybe there's a reason why it's like this and not like that". Coming to this repo and reading "Date was broken" is going to be of no help, and the JS world has a plethora of software artifacts on similar lines "X was broken, so I made my own". How was it broken? How are you keeping it simple and away from previous broken design? What are the surgical points that can be used for minimal invasion and maximum gain (e.g. like having the timezones, locales, non-gregorian calendars alone available in the browser) ?

I miss reading an obvious point regarding timezone support - evolvability. Yes, it's uncommon, but timezones do change. And then comes the question - which will update first? A library that may have a dozen instances part of the same deployment, or an execution environment? There's a net gain in saying: if tomorrow the US switches to one and only one timezone, users will just upgrade their browser (ever-green browsers are widespread now anyway) or their OS (also in the green-ish area) and all is good.

---

PS: @rxaviers thank you for the clear and well-argumented points you brought up. It has contributed to shortening my wall of text ;)

**Comment by ljharb** [comment]
@andreineculau Regarding padStart/padEnd, the "left-pad event" in fact showed that this stuff *does* need to be in the language - it did NOT in any way show that dependencies are "bad" or that things should be reinvented inline in each project (they shouldn't). Happy to educate anyone on that, but it's not relevant to this thread or this project, so let's please not debate it here.

Regarding "deprecating Date", that actually could be a viable option, but only once a proposal like this one lands - deprecating something without providing a replacement would be a terrible policy.

**Comment by andreineculau** [comment]
@ljharb 

> let's please not debate it here.

not at all my intention. I have pointed out that the pad-start-end proposal didn't have a rationale but a *emotionale*, similarly to this one. I'm talking strictly what I read in the text. Anything else is irrelevant. (FWIW I do not subscribe to your view towards the left-pad event)

> deprecating Date

I wrote that the first action point **of this temporal proposal** could have been to deprecate Date, not to have a standalone proposal with one action point: to deprecate Date.

**Comment by ljharb** [comment]
a) emotional reasons for things are just as important as "rational" ones; b) "X is difficult and Y makes it easier" is pretty clearly rational.

**Comment by andreineculau** [comment]
> emotional reasons for things are just as important as "rational" ones;

may be.

But the "rational" ones (I don't know why you used quotes) are completely missing.
And the emotional ones is only one: *Date has been a long time pain point in ECMAScript.*

PS: if hope you were not still talking about pad-start-end. We agreed we're not debating the padding proposal.

**Comment by maggiepint** [comment]
@rxaviers we talked at TC39. Are you good on motivations or do you want me to extrapolate more?

**Comment by mj1856** [comment]
@andreineculau, I'll encourage you to exercise patience.  We can update the motivations section in the readme at any time. There is no rush, and thus no need to go on the offensive with regard to the merits of rational vs. emotional arguments.  Thanks.

**Comment by dmitriid** [comment]
I'll chime in. @maggiepint  For some reason you prefer to write stuff in blog posts on your personal website (that you link to from within issues), and not include any of those in the proposal.

For example, the whole "the Date object in JS is broken in the following ways [here](https://maggiepint.com/2017/04/09/fixing-javascript-date-getting-started/) should really be in the motivation section of the proposal.

**Comment by mj1856** [comment]
@dmitriid - also, patience.

Also, there's a ton of material already written about the `Date` object's shortcomings.  We certainly *don't* want to hash that out all over again here.  That can only detract from this proposal.  Blogging and other activities are certainly preferred.

**Comment by ljharb** [comment]
Also, it's utterly sufficient to link to material from readmes; in no way does anything have to be inline.

**Comment by maggiepint** [comment]
Blog post is now linked from the readme.

**Comment by rxaviers** [comment]
@maggiepint, my questions above are still unanswered (or, at least, I didn't find the answers). Thanks

**Comment by dmitriid** [comment]
> @dmitriid - also, patience.

Patience for what? ;)

> Blog post is now linked from the readme.

Blog post basically shows that most obvious problems (such as timezones) can be solved with fixing the Date object. Some other (such as non-Gregorian calendars) don't have to be built-in, as they are too specific, and can most likely be implemented as libraries (basically, these issues have been raised by @rxaviers )

---

Dates are a pain and are broken basically everywhere. However, this is rarely the basis for introducing new things into the language. The basis and the motivation is specifying and fixing things that are actually broken, or need to be fixed, improved etc.

These descriptions and motivations (and also overviews of downsides and potential issues) are lacking across the board for most (all?) proposals in TC39, by the way.

Consider Java's [JSR](https://jcp.org/en/jsr/detail?id=310) for the new DateTime API. There's a separate overview _and_ whole sections titled "What need of the Java community will be addressed by the proposed specification?" and "Why isn't this need met by existing specifications".

**Comment by maggiepint** [comment]
I guess I'm not entirely sure how to respond to this entire thread. There
are extensive blog posts written by me and others on the problems with
date. I have linked to mine, others can be found with Google.

Yes, almost all of this can be done in a library. That doesn't mean we
shouldn't put it in ECMAscript. Show me the substantial portion of the
community who thinks this should remain in a library and I'll withdraw the
proposal.

**Comment by dmitriid** [comment]
> Yes, almost all of this can be done in a library. That doesn't mean we shouldn't put it in ECMAscript. 

No. This cannot be the argument. The argument should be: "these features have to be in ECMAScript because" and "these features have to be in a separate entity because".

- "left-pad can be implemented in a library.  That doesn't mean we shouldn't put it in ECMAScript"
- "React  can be implemented in a library.  That doesn't mean we shouldn't put it in ECMAScript"
etc.

For example:

There's a clear need for timezones to be handled "out of the box". Because timezone database is huge and we really don't want to download it on every request. Should it be implemented as `TimezonedInstant` or should `Date` be extended with it?

Non-gregorian calendars are not handled by `Date`. Should they really be built-in to the language? Why?

Coming back to JSR. There have been countless blog posts discussing how bad `java.util.date` and `java.calendar` are. And still the people proposing a replacement wrote *why* Java needed a replacement and *what actual needs* the replacement solves.

Timezones is a problem in search of a solution.
Non-gregorian calendars is a solution in need of a problem.
There are other things that are in between the two.

**Comment by rxaviers** [comment]
@maggiepint my questions above are pretty specific and based on your blog post. Please, just let me know if you have any questions and I'll be happy to clarify.

This thread is indeed getting pretty long, which shows the motivations aren't cristal clear to the community, which by the way is a basic part of a proposal. Having said that, I think this proposal has big potential and clarifying this point will benefit it.

**Comment by ljharb** [comment]
(To clarify: it shows they're not crystal clear to 5 of the 9 participants on the thread; five people is not sufficient to claim that it's unclear to the community)

**Comment by timrwood** [comment]
I took an hour or so to list some of the shortcomings of the existing `Date` object. 

#### Years and days are 1 indexed, but months, hours, minutes, and seconds are 0 indexed.

It is more common to treat months as 1 indexed, both in other programming languages and in human interactions.

#### The Date constructor does unexpected things depending on what arguments are passed in.

```js
new Date("0").toISOString()     // 2000-01-01T06:00:00.000Z
new Date(0).toISOString()       // 1970-01-01T00:00:00.000Z
new Date(0, 0).toISOString()    // 1900-01-01T06:00:00.000Z
new Date(0, 0, 0).toISOString() // 1899-12-31T06:00:00.000Z
```

#### The 2 digit year support is problematic.

```js
new Date( 99, 0, 1).toISOString() // 1999-01-01T06:00:00.000Z
new Date(100, 0, 1).toISOString() // 0100-01-01T06:00:00.000Z
new Date( 99, 0, 1).getYear() // 99
new Date(100, 0, 1).getYear() // -1800
```

#### Calculations are primarily done in system time, causing lots of strange bugs around DST.

```js
// After changing system time to America/Sao_Paulo
// On OSX: `sudo systemsetup settimezone America/Sao_Paulo`
new Date(2015, 9, 18).getDate() // 17 or 18 depending on the browser
```

#### Working with UTC time zone is more complicated than working with the system time zone.

```js
// system time
var a = new Date(2017, 0, 1)
a.getDate() // 1
a.getMonth() // 0
a.getFullYear() // 2017
// UTC time
var b = new Date(Date.UTC(2017, 0, 1))
a.getUTCDate() // 1
a.getUTCMonth() // 0
a.getUTCFullYear() // 2017
```

#### Working with a time zone other than UTC or system time is unsupported, and requires libraries whose data needs to be updated many times a year.

#### Adding and subtracting is not well supported. Using `Date#setX` methods often yield unexpected results.

```js
var a = new Date(2004, 1, 29)
a.setFullYear(2003)
a.toISOString() // 2003-03-01T03:00:00.000Z
var b = new Date(2004, 2, 31)
b.setMonth(3)
b.toISOString() // 2004-05-01T03:00:00.000Z
```

This list is not conclusive. As @mj1856 said, there are many other shortcomings that have been discussed elsewhere. I don't think enumerating them here should block progress on this proposal.

@maggiepint and @mj1856 maintain the most popular JS date library, have both given conference talks about time, and I strongly trust their opinions. They certainly have more domain knowledge than I do. I suggest that we trust that they know what they are doing and allow them to continue their work, adding to the motivations section as they see fit.

**Comment by ljharb** [comment]
(fwiw tho, months and days being zero-indexed actually makes perfect sense, since those are the only values you'd be highly likely to use as the index in an array of names, like Monday/Tuesday/etc, January/February/etc)

**Comment by mj1856** [comment]
When an array of month names is needed, is a smell that a formatting or parsing function is missing.  That work is in Ecma-402, and we'll leverage it.

Only bugs come from 0-based months.

**Comment by mj1856** [comment]
Also, looking up the month in an index is the wrong thing to optimize for.  Sure, if you have zero-based months you don't have to subtract one in an array index, but you'll still have to +/- 1 anywhere you construct a date from parts, such as business logic, user input, field validation, etc.

Common example:
```js
if (new Date() > new Date(2018, 1, 1)) {  // oops!
    alert('Happy New Year!');
}
```

**Comment by andreineculau** [comment]
> (To clarify: it shows they're not crystal clear to 5 of the 9 participants on the thread; five people is not sufficient to claim that it's unclear to the community)

@ljharb this is not a popularity context afaiac. @rxaviers pointed out the "thread is indeed getting pretty long", which is a VERY different thing, unless you think this thread is pure rambling and thus length equates to random characters being typed, not eloquent ideas and thoughts being communicated. To each their own, I guess.

> @maggiepint and @mj1856 maintain the most popular JS date library, have both given conference talks about time, and I strongly trust their opinions. They certainly have more domain knowledge than I do. I suggest that we trust that they know what they are doing and allow them to continue their work, adding to the motivations section as they see fit.

@timrwood I personally fully respect your opinion. 100%. Except that it cannot be an argument not to describe a problem domain. A solution cannot exist without a problem. Take for instance a super-simple HTTP header - [Sunset](https://tools.ietf.org/html/draft-wilde-sunset-header-03). If you go through the diffs of the previous versions, you'll notice how Erik develops predominantly the Introduction section, which is the motivation. That is a consistent pattern in the RFCs that I follow, and it comes as no surprise because once a proposal is put forward, even the author(s) remodel their take, not just on the proposed solution, but on what consists a problem and how to reason about the problem.

Even I, some anonymous individual, can come up with a solution to the generic "pain point" in javascript's Date. In a zippy even. I might even have extensive knowledge about the domain. But the bare minimum is that I can still put down at least the same amount of information that you have in your comment, for the 100-th time after blog posts and talks, so that others can read top-down, left-to-right, *everything* that is wrong today, and is *not* to be part in the solution. To put it differently, you can be <insert your idol here>, but if you cannot or don't want to tell others the answer to "so what?", then others will both not be able to confirm you are right, to point out what is wrong or what can be improved, nor possibly choose your solution over the problematic situation itself.

Perfect [example with the zero-index](https://github.com/tc39/proposal-temporal/issues/19#issuecomment-320776634) . For reasons x, y, z zero-index is considered problematic in this proposal and is the *motivation* to use "real"-index across the new proposal. *That* is a sentence that one can follow and judge.

Saying that I can Google this stuff up assumes so many things that devalues the process beyond recognition. So I should take at face-value everything that pops up in my *personalized* google results page about JavaScript's Date? And what if Google goes bust? or your blog goes bust? or if conferences take down your videos? Ok, nothing goes bust and I do find all the right google results. So now everyone has to google, read random blog posts, watch random videos, thus have a super sparse baseline and provide consistent feedback? Isn't it better for 1-2-3 people to actually take 1-2-3 hours, even days and nail the problem domain so that hundreds, thousands, millions of other people can just read everything in 15 minutes?

Patience is virtue that I randomly have, but the invitation to patience without a promise of fulfillment is disrespectful. [Logic is also a virtue by the way.](https://culbreath.wordpress.com/2012/09/10/logic-and-virtue/)

---

Related read: [The So What Question](http://blog.writingpower.net/2008/04/22/the-so-what-question/).

**Comment by mj1856** [comment]
@andreineculau Please review [the TC39 process](https://tc39.github.io/process-document/).  This proposal is currently in stage 1.  The committee approved (at the [March 2017 meeting](https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-03/summary.md)) entrance to stage 1, and thus was satisfied that the case for the proposal (point 1 under stage 1 purpose) and the prose outlineing the problem (point 2 under stage 1 entrance criteria) have been met.

The detailed rationale that you are asking for is not required at this stage, but will before reaching stage 3 (Spec quality complete, all semantics described, etc.)

This is why I asked for patience.  We don't necessarily disagree with you on all points - but you seem to be saying that nothing should be shown here at all yet - and that's just not how this works.  We (@maggiepint, @bterlson, and myself) are championing this proposal through the process, and we are doing so in the open.  What you see in the readme is our work in progress.  It's not even a stage-2 draft yet.

**Comment by andreineculau** [comment]
@mj1856 I'm well aware of the theoretical process. Please correct me if I'm wrong, but it reads that "prose outlining the problem" is part of Stage 1 acceptance. Everything else that follows is related to the spec. Looking at the staged process, I actually cannot use the process document "against" you, to require you to improve the motivation. Examining the problem space should have happened before stage 1. And by all means, I know at least a bunch of the `Date` quirks, and I have read online material, talks from you and others, so by all means, this is not undermining the energy and time devoted to examining the problem space. But at face-value, after all this time and energy, when it came to writing a proposal, all that was transposed into prose was "Date has been a long time pain point in ECMAScript.".

> you seem to be saying that nothing should be shown here at all yet

Not at all, and I'm willing to bet the others are not either. Far from in fact. I mean, forget even the TC39 process. If that was what I'm saying, I don't think I would have even left a comment. If this was not a work in progress, the same.

In all honesty, I don't know how the textual proposal made it to Stage 1. I mean no offense to anyone. I swear. It's just that the content itself describes nothing of the problem space. It was probably that there's a general truth that goes `Date is way too broken, I won't even start telling you how broken`. Or we know the background work of the x, y, z, we trust them blindly, let's move on. Content-wise there is nothing to say that `Date` actually has to be fixed and where (i.e. which are the weak points, and which are the strong points that should not be touched).

For all that it's worth, you're now saying that

* you agree with me/others on "some" points (which? I don't know. But I know: patience),
* and that "these" points will be addressed before stage 3 (why? I don't know. Like I said, the problem space is to be covered in its majority, not entirety, pre-stage-1 as far as the process document goes)

I can wait. Not a problem, although

> We certainly don't want to hash that out all over again here. That can only detract from this proposal.

I 100% do not understand how being meticulous about clarifying the problem space makes a solution less worthy. I guess research papers could be so much better with no problem space and reference whatsoever.

> Blogging and other activities are certainly preferred.

I 100% do not understand how that is ok because <reasons in a previous comment>.

> I guess I'm not entirely sure how to respond to this entire thread. There are extensive blog posts written by me and others on the problems with date. I have linked to mine, others can be found with Google.

@maggiepint also subscribes to the idea to not actually do anything. At least it reads that she doesn't see a problem in this thread, thus a course of action is non-existant.

And last, but not least - @rxaviers still has his spot-on questions unanswered.

---

*I said everything I had to say on this topic, in much more words and with much more patience that I had ever imagined. From here on, no words from me, just eternal patience. Thanks, bonne chance!*

**Comment by maggiepint** [comment]
Locked. Will respond to Ryan's points later.

**Comment by bterlson** [comment]
In the future let's work to improve the tone of these discussions. While making no claim about anyone's intent, many of the posts above are needlessly combative and hostile. 

TC39's standards process does not require strong or detailed motivations for stage 1 proposals, merely consensus that some kind of investigation of the problem space is warranted for whatever reason. This is by design. There is no point in arguing about what TC39's process is with TC39 delegates; we know what we're doing. Please do not use this repository to argue for changes to our process.

The champions are well aware that some would like to see detailed documentation of the motivations behind this proposal. A simple request for more detail would have sufficed.

Additionally it is clear that there are some concerns about this proposal and alternative proposals out there (including doing nothing). These concerns and proposals should be discussed in their own issues where actual substantive discussions can take place over the coming weeks and months.
